src/components/deal-proposal-form/DealProposalForm.tsx

```diff
@@ -0,0 +1,223 @@
+import { zodResolver } from "@hookform/resolvers/zod";
+import type { InjectedAccountWithMeta } from "@polkadot/extension-inject/types";
+import { useMemo } from "react";
+import { useForm } from "react-hook-form";
+import { Toaster } from "react-hot-toast";
+import { z } from "zod";
+import { blockToTime, planckToDot } from "../../lib/conversion";
+import { type StorageProviderInfo, isStorageProviderInfo } from "../../lib/storageProvider";
+import { HookAccountSelector } from "./AccountSelector";
+import { DisabledInputInfo } from "./DisabledInputInfo";
+import { HookInput } from "./Input";
+import { PieceUploader } from "./PieceUploader";
+import { ProviderSelector } from "./ProviderSelector";
+import type { IFormValues } from "./types";
+
+const BLOCKS_IN_MINUTE = 10;
+const OFFSET = BLOCKS_IN_MINUTE * 5;
+const DEFAULT_DEAL_DURATION = 50;
+const DEFAULT_MAX_PROVE_COMMIT_DURATION = 50;
+const DEFAULT_MAX_DEAL_DURATION = 180 * BLOCKS_IN_MINUTE;
+
+// TODO(@th7nder,16/04/2025): fetch from chain
+// This is the minimum amount of blocks it'll take for the deal to be active.
+const maxProveCommitDuration = DEFAULT_MAX_PROVE_COMMIT_DURATION;
+// It's not in pallet metadata anymore, because of the benchmarks.
+const minDealDuration = DEFAULT_DEAL_DURATION;
+const maxDealDuration = DEFAULT_MAX_DEAL_DURATION;
+
+const storageProviderInfoSchema = z.custom<StorageProviderInfo>(isStorageProviderInfo);
+
+function validationSchema(currentBlock: number) {
+  return z
+    .object({
+      piece: z.object({
+        pieceCid: z.string(),
+        payloadCid: z.string(),
+        size: z.number(),
+        file: z.instanceof(File).refine((file: File) => file.size <= 8 * 1024 * 1024, {
+          message: "File must be under 8MB",
+        }),
+      }),
+      label: z.string().max(128),
+      startBlock: z.coerce
+        .number()
+        .int()
+        .positive()
+        .refine((value) => value > currentBlock + maxProveCommitDuration, {
+          message: `Must be at least ${maxProveCommitDuration} blocks in the future`,
+        }),
+      endBlock: z.coerce.number().int().positive(),
+      pricePerBlock: z.coerce.number().int().positive(),
+      providerCollateral: z.coerce.number().int().positive(),
+      client: z.string(),
+      providers: z
+        .array(storageProviderInfoSchema)
+        .min(1, "There must be at least 1 Storage Provider selected"),
```

Comment:
Is there no way to synthesize an event that triggers this before the user clicks continue? I still find that issue annoying

Reply:
I made it's validated on every change (to the field).